src/lib.rs

```diff
@@ -169,3 +178,46 @@ impl DeviceManager {
         Ok(())
     }
 }
+
+async fn get_device_id(opt_device_id: Option<String>) -> Result<String, DeviceManagerError> {
+    if let Some(device_id) = opt_device_id {
+        Ok(device_id)
+    } else {
+        let connection = zbus::Connection::system().await?;
+        let proxy = DeviceProxy::new(&connection).await?;
+        let hardware_id: String = proxy.hardware_id().await?;
+        if hardware_id.is_empty() {
+            return Err(DeviceManagerError::FatalError(
+                "No hardware id provided".to_string(),
+            ));
+        }
+        Ok(hardware_id)
+    }
+}
+
+async fn get_credentials_secret(
+    device_id: &str,
+    opts: &DeviceManagerOptions,
+) -> Result<String, DeviceManagerError> {
+    if let Some(secret) = opts.credentials_secret.clone() {
+        Ok(secret)
+    } else if Path::new(&format!("./{}.json", device_id)).exists() {
+        let reader = File::open(&format!("./{}.json", device_id)).unwrap();
+        Ok(serde_json::from_reader(reader).expect("Unable to read secret"))
+    } else if let Some(token) = opts.pairing_token.clone() {
+        let registration =
+            registration::register_device(&token, &opts.pairing_url, &opts.realm, &device_id).await;
+        match registration {
+            Err(_e) => return Err(DeviceManagerError::FatalError("Pairing error".to_string())),
```

Comment:
you can map the error and then `?` it, instead of doing a whole `match`

Reply:
the PairingError is not publicly exposed by the SDK